Preserve custom ops via run_decomps

tugsbayasgalan opened the pull request:
Summary: Previously, our flow only works for aten ops but there are some users are passing in custom ops to decomp table. So we want to also support that.

Test Plan: Executorch tests

Differential Revision: D63470928

bdhirsh commented on torch/_decomp/__init__.py:
```diff
@@ -299,6 +302,10 @@ def _is_cia_op(op: "OpOverload") -> bool:
 
 
+def _is_preservable_cia_op(op: "OperatorBase") -> bool:
+    return _check_valid_to_preserve(op) and _is_cia_op(op)
```
tbh, `_collect_all_valid_cia_ops` and its related helpers could probably all go in `exported_program.py` instead of needing to live here with the other decomp main API's

tugsbayasgalan replied:
Yeah hahha good idea! I will move them.

tugsbayasgalan replied:
Hmm actually i think it is bit hard because we do need this APIs to create the new core_aten_decomp table. Can I follow up on it later after I move all core_aten_decomp references to export alias of core_aten_decomp? Then we will be doing this compute in exported_program.py.

bdhirsh commented on torch/_decomp/__init__.py:
```diff
@@ -40,9 +40,9 @@
-global_decomposition_table: Dict[
-    str, Dict[torch._ops.OperatorBase, Callable]
-] = defaultdict(dict)
+global_decomposition_table: Dict[str, Dict[torch._ops.OperatorBase, Callable]] = (
```
was this file not linted properly before?

tugsbayasgalan replied:
I think i was using linter from fbcode but it is not good

tugsbayasgalan committed:
Preserve custom ops via run_decomps (#136773)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/136773

Previously, our flow only works for aten ops but there are some users are passing in custom ops to decomp table. So we want to also support that.

Test Plan: Executorch tests

Differential Revision: D63470928
